### PR TITLE
18024 Incremental AGM extension work

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#18024

**This is some up-front architecture/layout work as I figure out how to handle the data and logic for this filing.**

*Description of changes:*
- app version = 6.9.5
- imported shared DatePicker component
- renamed a base class
- added a base class
- saved user info into store
- added initial AboutTheBusiness sub-component
- added initial AgmExtensionEvaluation sub-component
- added AgmExtensionHelp sub-component
- added initial ExtensionRequest sub-component
- added VcardTemplate
- updated tooltip text
- updated FF
- added AgmExtEval interface
- renamed RootStateIF -> AuthenticationStateIF
- renamed StateIF -> RootStateIF
- added DateMixin function for DatePicker component
- updated fetchUserInfo to return object (not Axios response)
- added missing default FF
- deleted obsolete dialog, transition, buttons, methods, CSS from AGM Extension
- added help section
- added initial sub-components to AGM Extension
- updated intro text
- updated some whitespace
- updated filing code
- removed delay in mounted() in AGM Extension
- updated onHaveChanges() watcher in AGM Extension
- pre-populated CertifiedBy
- deleted obsolete transition, methods, CSS from AGM Extension
- removed delay in mounted() in AGM Location Chg
- updated onHaveChanges() watcher in AGM Location Chg
- updated unit tests
- misc cleanup

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

File: src/components/common/VcardTemplate.vue
Comment (severinbeauvais) [inline src/components/common/VcardTemplate.vue:1]: This is a template to display the Vcard sections consistently and easily, ie:

image

In a future PR, we might want to pass in a validation prop to show the content div with error styling.

**Review by leodube-aot — APPROVED** [review]

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Comment by severinbeauvais** [comment]
If the rest of you could review this, please...

My intent with this was to prepare the layout and architecture of the AGM Extension filings so that the remaining tickets can be done by multiple people. The main thing I wasn't sure about was passing data between the parent component and the sub-components. It turned out to be very easy -- let me know if you want a walk-through.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Review by chenhongjing — APPROVED** [review]
LGTM 😺

File: src/components/AgmExtension/ExtensionRequest.vue [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            sm="3"
+          >
+            <strong>Is this the first AGM?</strong>
+          </v-col>
+          <v-col
+            cols="12"
+            sm="9"
+          >
+            <v-radio-group
+              v-model="data.isFirstAgm"
```
Comment (severinbeauvais) [inline src/components/AgmExtension/ExtensionRequest.vue:25]: The v-model is a 2-way binding.

Since `data` was passed by reference, updating "isFirstAgm" here updates it in the data object.

File: src/components/AgmExtension/ExtensionRequest.vue [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    DatePicker,
+    VcardTemplate
+  }
+})
+export default class ExtensionRequest extends Vue {
+  @Prop({ required: true }) readonly data!: AgmExtEvalIF
+
+  @Emit('data')
+  private emitData () {
+    return this.data
```
Comment (severinbeauvais) [inline src/components/AgmExtension/ExtensionRequest.vue:134]: This is not used yet but should allow emitting the data object back to the parent component.

File: src/interfaces/agm-ext-eval-interface.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  incorporationDate: Date
+  isFirstAgm: boolean
+  agmYear: string // YYYY
+  prevAgmDate?: string // YYYY-MM-DD
+  isPrevExtension?: boolean
+  prevExpiryDate?: string // YYYY-MM-DD
+  intendedAgmDate?: string // YYYY-MM-DD
+  extensionDuration: number // in months
+  agmDueDate: string // YYYY-MM-DD
+}
```
Comment (severinbeauvais) [inline src/interfaces/agm-ext-eval-interface.ts:13]: This is my first pass at this. It can be changed.

File: src/mixins/date-mixin.ts [hunk 4]
```diff
@@ -93,6 +93,16 @@ export default class DateMixin extends Vue {
...
+  /**
+   * Converts a date string (YYYY-MM-DD) to a date string (Month Day, Year) in Pacific timezone.
+   * @param longMonth whether to show long month name (eg, December vs Dec)
+   * @param showWeekday whether to show the weekday name (eg, Thursday)
+   * @example "2021-01-01" -> "Thursday, December 31, 2020"
+   */
+  yyyyMmDdToPacificDate (dateStr: IsoDatePacific, longMonth = false, showWeekday = false): string {
+    return DateUtilities.dateToPacificDate(this.yyyyMmDdToDate(dateStr), longMonth, showWeekday)
+  }
+
```
Comment (severinbeauvais) [inline src/mixins/date-mixin.ts:105]: This is needed by the shared DatePicker component.

File: src/views/AgmLocationChg.vue [hunk 5]
```diff
@@ -14,30 +14,8 @@
...
-          </div>
-        </div>
-      </div>
-    </v-fade-transition>
-
     <!-- Main Body -->
-    <v-container
-      v-if="dataLoaded"
-      class="view-container"
-    >
```
Comment (severinbeauvais) [inline src/views/AgmLocationChg.vue:40]: @JazzarKarim Since we don't fetch drafts (or load any extra data), we don't need the loading spinner or a conditional on the main body.

File: src/views/AgmLocationChg.vue [hunk 6]
```diff
@@ -441,12 +408,11 @@ export default class AgmLocationChg extends Mixins(CommonMixin, DateMixin,
...
 
     // this is the id of THIS filing
-    // if 0, this is a new filing
-    // otherwise it's a draft filing
+    // it must be 0 (meaning new filing) -- we do not support resuming a draft filing
     this.filingId = +this.$route.params.filingId // number or NaN
 
     // if required data isn't set, go back to dashboard
-    if (!this.getIdentifier || isNaN(this.filingId)) {
+    if (!this.getIdentifier || this.filingId !== 0) {
```
Comment (severinbeauvais) [inline src/views/AgmLocationChg.vue:415]: Checking the filing ID, which we pass as 0 in the router push call, ensures that the user cannot go to the URL manually.

```
this.$router.push({ name: Routes.AGM_LOCATION_CHANGE, params: { filingId: '0' } })
```

File: src/views/AgmExtension.vue [hunk 7]
```diff
@@ -73,175 +45,42 @@
...
-                      (radio buttons go here)
-                    </v-col>
-                  </v-row>
-                </v-card>
-              </div>
-            </section>
+            <!-- About the Business -->
+            <AboutTheBusiness
+              class="mt-8"
+              :data.sync="data"
```
Comment (severinbeauvais) [inline src/views/AgmExtension.vue:64]: This should allow 2-way sync of the data object.

File: src/views/AgmExtension.vue [hunk 8]
```diff
@@ -393,62 +217,54 @@ export default class AgmExtension extends Mixins(CommonMixin, DateMixin,
...
   @Getter(useBusinessStore) getLegalName!: string
   @Getter(useConfigurationStore) getPayApiUrl!: string
+  @Getter(useBusinessStore) isGoodStanding!: boolean
   @Getter(useRootStore) isRoleStaff!: boolean
 
   // enum for template
   readonly FilingCodes = FilingCodes
 
+  // evaluation object
+  data = { ...EmptyAgmExtEval } as AgmExtEvalIF
```
Comment (severinbeauvais) [inline src/views/AgmExtension.vue:227]: This initializes the data object with empty fields.

File: src/components/AgmExtension/ExtensionRequest.vue (hunk 2 above)
Comment (severinbeauvais) [reply src/components/AgmExtension/ExtensionRequest.vue:134]: Obsolete comment.

File: src/components/AgmExtension/ExtensionRequest.vue [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            sm="3"
+          >
+            <strong>AGM Year</strong>
+          </v-col>
+          <v-col
+            cols="12"
+            sm="9"
+          >
+            <v-text-field
+              v-model="agmYear"
```
Comment (severinbeauvais) [inline src/components/AgmExtension/ExtensionRequest.vue:52]: Here's another way to enable data sync with the parent component.

Here, use what looks like a local variable....

File: src/views/AgmLocationChg.vue (hunk 5 above)
Comment (JazzarKarim) [reply src/views/AgmLocationChg.vue:40]: You're absolutely right. Good catch Sev. I probably should've made this change in one of my PRs.

File: src/components/AgmExtension/ExtensionRequest.vue [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  get agmYear (): string {
+    return this.data.agmYear
+  }
+
+  set agmYear (value: string) {
+    this.data.agmYear = value
+    this.emitData()
+  }
```
Comment (severinbeauvais) [inline src/components/AgmExtension/ExtensionRequest.vue:144]: ... But actually, `agmYear` is both a getter (from data prop) and a setter, which emits/syncs the object back to the parent component.

In practise, the other approach is simpler (see `isFirstAgm`), but since this component will also contain logic and compute data, emitting/syncing will also be needed.

File: src/views/AgmExtension.vue [hunk 11]
```diff
@@ -73,175 +45,42 @@
...
+            />
+
+            <!-- AGM Extension Evaluation -->
+            <AgmExtensionEvaluation
+              class="mt-8"
+              :data.sync="data"
+            />
+
+            <!-- delete this debugging code -->
+            <pre class="mt-8">data: {{ data }}</pre>
```
Comment (severinbeauvais) [inline src/views/AgmExtension.vue:80]: For debugging. A future PR can delete this if unneeded.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

File: src/components/AgmExtension/ExtensionRequest.vue (hunk 9 above)
Comment (severinbeauvais) [reply src/components/AgmExtension/ExtensionRequest.vue]: Obsolete. Resolving.

File: src/components/AgmExtension/ExtensionRequest.vue [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+          <v-col
+            cols="12"
+            sm="9"
+          >
+            <v-text-field
+              class="agm-year-textfield"
+              filled
+              label="AGM Year"
+              :value="agmYear"
+              @change="agmYear = $event"
```
Comment (severinbeauvais) [inline src/components/AgmExtension/ExtensionRequest.vue:56]: This is another way to handle data passing...

Here, I pass in `value` and capture the output on the `change` event. ("Input" triggers on every keystroke while "change" triggers when the user has finished entering data. I used "change" to only set the data once the user has fully entered it and to reduce the number of syncs.)

See "agmYear" in the script below.

**Comment by severinbeauvais** [comment]
If you have previously reviewed this, please look over my last few comments.

If you havent reviewed this yet, please do - thanks :)

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Review by ketaki-deodhar — COMMENTED** [review]
Looks good Sev! Your detailed comments help understand the logic behind the written code :)

**Review by ketaki-deodhar — APPROVED** [review]

File: src/views/AgmLocationChg.vue [hunk 13]
```diff
@@ -642,10 +638,12 @@ export default class AgmLocationChg extends Mixins(CommonMixin, DateMixin,
...
-  @Watch('agmYearValid')
-  @Watch('agmLocationValid')
-  @Watch('certifyFormValid')
-  @Watch('resonValid')
+  /** Watches all data properties to keep track of changes. */
+  @Watch('agmYear')
+  @Watch('agmLocation')
+  @Watch('certifiedBy')
+  @Watch('isCertified')
+  @Watch('reason')
```
Comment (JazzarKarim) [inline src/views/AgmLocationChg.vue:646]: Why did you change these Sev?

Comment (severinbeauvais) [reply src/views/AgmLocationChg.vue:646]: To directly check if the user has changed any input fields.

Comment (severinbeauvais) [reply src/views/AgmLocationChg.vue:646]: For example, if the user entered certifiedBy but didn't check isCertified, the certify form would be invalid, but they have in fact entered some data.

Comment (JazzarKarim) [reply src/views/AgmLocationChg.vue:646]: Got it, thanks! Resolving.

File: src/components/AgmExtension/AgmExtensionHelp.vue [hunk 14]
```diff
@@ -0,0 +1,31 @@
+<template>
+  <section class="agm-location-help mt-0">
```
[the remark this answers is not in the document]
Comment (severinbeauvais) [reply src/components/AgmExtension/AgmExtensionHelp.vue]: Fixed.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app

**Review by JazzarKarim — APPROVED** [review]
Looks great Sev! 👍 This is super helpful in order for multiple developers to dig in at the same time if needed.

**Comment by severinbeauvais** [comment]
Do not merge until Business Digital Credentials is out of Dev environment.

**Comment by JazzarKarim** [comment]
Sev, would you like to merge before me? I don't mind doing the rebase if you would like.

**Comment by severinbeauvais** [comment]
> Sev, do you want to merge before me? I don't mind doing the rebase if you would like.

I've made some extra changes on top of what you did. You sure you don't mind rebasing?

**Comment by JazzarKarim** [comment]
> > Sev, do you want to merge before me? I don't mind doing the rebase if you would like.
> 
> I've made some extra changes on top of what you did. You sure you don't mind rebasing?

Actually Sev, I'm just going to merge mine now. I think it'll be easier for you to rebase than me. Is that OK with you?

**Comment by severinbeauvais** [comment]
> Actually Sev, I'm just going to merge mine now. I think it'll be easier for you to rebase than me. Is that OK with you?

Do it.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-filings-dev--pr-558-eyhk80mt.web.app
